src/approx_binarytree_itensornetwork.jl

```diff
@@ -0,0 +1,512 @@
+  |  2
+  9 /
+  |/
+  1
+
+  The density matrix for the root 3, children 5 squares the subgraph
+    with vertices 5. and contract with the tensor 3
+    |
+    3
+   /
+  5 - 5 -
+"""
+struct _PartialDensityMatrix
+  tensor::ITensor
+  root::Union{<:Number,Tuple}
+  child::Union{<:Number,Tuple}
+end
+
+"""
+The struct contains cached density matrices and cached partial density matrices
+for each vertex in the tensor network.
+"""
+struct _DensityMatrixAlgCaches
+  v_to_cdm::Dict{Union{<:Number,Tuple},_DensityMatrix}
+  v_to_cpdms::Dict{Union{<:Number,Tuple},Vector{_PartialDensityMatrix}}
+end
+
+function _DensityMatrixAlgCaches()
+  v_to_cdm = Dict{Union{<:Number,Tuple},_DensityMatrix}()
+  v_to_cpdms = Dict{Union{<:Number,Tuple},Vector{_PartialDensityMatrix}}()
+  return _DensityMatrixAlgCaches(v_to_cdm, v_to_cpdms)
+end
+
+"""
+Remove cached partial density matrices from `cpdms` whose child is in `children`
+"""
+function _remove_cpdms(cpdms::Vector, children)
+  return filter(pdm -> !(pdm.child in children), cpdms)
+end
+
+"""
+The struct stores data used in the density matrix algorithm.
+  partition: The given tn partition
+  out_tree: the binary tree structure of the output ITensorNetwork
+  root: root vertex of the bfs_tree for truncation
+  innerinds_to_sim: mapping each inner index of the tn represented by `partition` to a sim index
+  caches: all the cached density matrices
+"""
+struct _DensityMartrixAlgGraph
+  partition::DataGraph
+  out_tree::NamedGraph
+  root::Union{<:Number,Tuple}
+  innerinds_to_sim::Dict{<:Index,<:Index}
+  caches::_DensityMatrixAlgCaches
+end
+
+function _DensityMartrixAlgGraph(
+  partition::DataGraph, out_tree::NamedGraph, root::Union{<:Number,Tuple}
+)
+  innerinds = _get_inner_inds(partition)
+  sim_innerinds = [sim(ind) for ind in innerinds]
+  return _DensityMartrixAlgGraph(
+    partition,
+    out_tree,
+    root,
+    Dict(zip(innerinds, sim_innerinds)),
+    _DensityMatrixAlgCaches(),
+  )
+end
+
+function _get_inner_inds(partition::DataGraph)
+  networks = [Vector{ITensor}(partition[v]) for v in vertices(partition)]
+  network = vcat(networks...)
+  outinds = noncommoninds(network...)
+  allinds = mapreduce(t -> [i for i in inds(t)], vcat, network)
+  return Vector(setdiff(allinds, outinds))
+end
+
+function _get_out_inds(partition::DataGraph)
+  networks = [Vector{ITensor}(partition[v]) for v in vertices(partition)]
+  network = vcat(networks...)
+  return noncommoninds(network...)
+end
+
+"""
+Contract of a vector of tensors, `network`, with a contraction sequence generated via sa_bipartite
+"""
+function _optcontract(network::Vector)
+  @timeit_debug ITensors.timer "[approx_binary_tree_itensornetwork]: _optcontract" begin
+    if length(network) == 0
+      return ITensor(1.0)
+    end
+    @assert network isa Vector{ITensor}
+    @timeit_debug ITensors.timer "[approx_binary_tree_itensornetwork]: contraction_sequence" begin
+      seq = contraction_sequence(network; alg="sa_bipartite")
```

Comment:
@mtfishman The reason I import `OMEinsumContractionOrders` is that I use a subroutine that automatically contract a vector of tensors with reasonable contraction sequence, and the one I'm using here is `sa_bipartite`. I'm not sure if we can have this without directly `using OMEinsumContractionOrders`

Reply:
A user would have to load `OMEinsumContractionOrders` before loading `ITensorNetworks` to use functionality that requires `OMEinsumContractionOrders` (such as your function `approx_binary_tree_itensornetwork`), i.e.:
```julia
using OMEinsumContractionOrders
using ITensorNetworks

tn = approx_binary_tree_itensornetwork(...)
```
so that we don't have to depend directly on `OMEinsumContractionOrders` in the library. It's a way to make it so that not everyone depends on certain libraries if they don't need the functionality that is enabled by that library.

I would consider depending directly on `OMEinsumContractionOrders` since it's not a heavy dependency but let's hold off on that decision for future development.

Reply:
that makes sense!

Reply:
We could also pass through a `contraction_sequence_alg` keyword argument from the top level function `approx_itensornetwork` so users could choose a backend (also `contraction_sequence_kwargs=(;)` which would be a `NamedTuple` storing keyword arguments to pass to `contraction_sequence` to customize even more).

Reply:
SG, done